pesudocode/voxel_grid_structure_3.hpp

```diff
@@ -0,0 +1,142 @@
+#include <iostream>
+#include <type_traits>
+#include <vector>
+#include <unordered_map>
+#include <Eigen/Dense>
+#include <optional>
+
+class PointXYZ
+{
+public:
+    PointXYZ() {}
+};
+
+template <typename PointT>
+class PointCloudT
+{
+public:
+    PointCloudT() {}
+
+    void push_back(PointT p) {}
+
+    // Defined in pcl::PointCloud
+    using PointType = PointT;
+};
+
+template <typename PointT>
+class Filter
+{
+    using PointCloud = PointCloudT<PointXYZ>;
+
+public:
+    Filter() {}
+    virtual void applyFilter(PointCloud &pc) = 0;
+    void filter(PointCloud &output) { applyFilter(output); }
+};
+
+// ================== Base of VoxelGrid ==================
+
+// Deduce type
+
+// I've tried deducing by: https://stackoverflow.com/questions/22632236/how-is-possible-to-deduce-function-argument-type-in-c
+// but it doesn't work in our case
+template <typename T>
+using get_point_type = typename T::PointCloud::PointType;
+
+template <typename T>
+using get_grid_type = typename T::Grid;
```

Comment:
These are good enough for now. In future, we'll use the function definitions (in the GridStruct) to deduce them